crates/ruff_linter/src/rules/numpy/rules/numpy_2_0_deprecation.rs

```diff
@@ -701,3 +707,233 @@ pub(crate) fn numpy_2_0_deprecation(checker: &mut Checker, expr: &Expr) {
+/// contain any [`ast::ExprAttribute`] nodes that indicate the numpy
+/// member is being accessed from the non-deprecated location?
+fn try_block_contains_undeprecated_attribute(
+    try_node: &ast::StmtTry,
+    replacement_details: &Details,
+    semantic: &SemanticModel,
+) -> bool {
+    let Details::AutoImport {
+        path,
+        name,
+        compatibility: _,
+    } = replacement_details
+    else {
+        return false;
+    };
+    let undeprecated_qualified_name = {
+        let mut builder = QualifiedNameBuilder::default();
+        for part in path.split('.') {
+            builder.push(part);
+        }
+        builder.push(name);
+        builder.build()
+    };
+    let mut attribute_searcher = AttributeSearcher::new(undeprecated_qualified_name, semantic);
+    for stmt in &try_node.body {
+        attribute_searcher.visit_stmt(stmt);
+        if attribute_searcher.found_attribute {
+            return true;
+        }
+    }
+    false
+}
+
+/// AST visitor that searches an AST tree for [`ast::ExprAttribute`] nodes
+/// that match a certain [`QualifiedName`].
+struct AttributeSearcher<'a> {
+    attribute_to_find: QualifiedName<'a>,
+    semantic: &'a SemanticModel<'a>,
+    found_attribute: bool,
+}
+
+impl<'a> AttributeSearcher<'a> {
+    fn new(attribute_to_find: QualifiedName<'a>, semantic: &'a SemanticModel<'a>) -> Self {
+        Self {
+            attribute_to_find,
+            semantic,
+            found_attribute: false,
+        }
+    }
+}
+
+impl Visitor<'_> for AttributeSearcher<'_> {
+    fn visit_expr(&mut self, expr: &'_ Expr) {
+        if self.found_attribute {
+            return;
+        }
+        if expr.is_attribute_expr()
+            && self
+                .semantic
+                .resolve_qualified_name(expr)
+                .is_some_and(|qualified_name| qualified_name == self.attribute_to_find)
+        {
+            self.found_attribute = true;
+            return;
+        }
+        ast::visitor::walk_expr(self, expr);
+    }
+}
+
+/// Given an [`ast::StmtTry`] node, does the `try` branch of that node
+/// contain any [`ast::StmtImportFrom`] nodes that indicate the numpy
+/// member is being imported from the non-deprecated location?
+fn try_block_contains_undeprecated_import(
+    try_node: &ast::StmtTry,
+    replacement_details: &Details,
+) -> bool {
+    let Details::AutoImport {
+        path,
+        name,
+        compatibility: _,
+    } = replacement_details
+    else {
+        return false;
+    };
+    let mut import_searcher = ImportSearcher::new(path, name);
+    for stmt in &try_node.body {
+        import_searcher.visit_stmt(stmt);
+        if import_searcher.found_import {
+            return true;
+        }
+    }
+    false
+}
+
+/// AST visitor that searches an AST tree for [`ast::StmtImportFrom`] nodes
+/// that match a certain [`QualifiedName`].
+struct ImportSearcher<'a> {
+    module: &'a str,
+    name: &'a str,
+    found_import: bool,
+}
```

Comment:
I don't think it matters much because we only run this code in the error case, but it is kind of expensive. Do we have no way of resolving the statement in which a binding is declared and then traverse upwards (by using the parent API?)

Reply:
The situation is that we're visiting an `except` block in a `try`/`except` statement. We find a deprecated numpy access in that `except` block. We need to find out if there's an undeprecated access in the `try` block of the `try`/`except` statement. The way I do this is by traversing upwards from the node we're linting in the `except` block (using the parent API) to find the `StmtTry` node, then from the `StmtTry` node I then traverse downwards into the `try` block using the `ImportSearcher` to find if there are any imports using the undeprecated API.